iree/compiler/Dialect/Flow/Transforms/ConvertToFlowTensorOps.cpp

```diff
@@ -250,6 +250,32 @@ struct SubTensorToTensorSlice : public OpRewritePattern<SubTensorOp> {
   }
 };
 
+/// Converts linalg.fill ops that are not used by other Linalg ops into
+/// flow.tensor.splat ops.
+///
+/// This is expected to improve performance because we can use DMA
+/// functionalities for the fill, instead of dispatching kernels.
+///
+/// It assumes linalg.fill ops that has linalg op users can be fused with
+/// its users so those cases aren't supported here.
+struct LinalgFillToFlowTensorSplat final
```

Comment:
Sorry for the repeated nag on this one, but adding it by default here (when run before DispatchLinalgOnTensors) will avoid any fusion of fill with other ops. That is better than always doing it only the whole buffer. My comment from previous time was that the same pass should be controlled via flag to run "before" or "after" dispatch region formation. All current patterns are to be run before. The fill can be made to run after.

Reply:
Having the pattern to run after the DispatchLinalgOnTensors pass would mean we need to first *identify* and *exclude* the standalone `linalg.fill` ops in DispatchLinalgOnTensors (because otherwise the `linalg.fill` op would be put into its own dispatch region like shown by the existing [test](https://github.com/google/iree/blob/2dfdc8d9c80195947ea847d79da5b5010ab23801/iree/compiler/Dialect/Flow/Transforms/test/dispatch_linalg_on_tensors.mlir#L324-L361)) to allow the ConvertToFlowTensorOpsAfterDispatch to kick in later. So that's coupling two passes and introducing quite a lot boilerplate code for just a canoncialization-like direct 1:1 op rewrite. Frankly I'm not so sure about that's better.

Reply:
No you just need to add a flag to the pass `isRunBeforeDispatchRegionFormation` . 

If `isRunBefore*` is true, use the patterns that exist in that pass.
If `isRunBefore*` is false, use the pattern to convert `fill` to `flow.*`.

Its essentially two passes in one, but it will collect all the conversion to `flow.*` ops in one-place.

The the pass-pipeline in Flow is

```
convertToFlowTensorOps(/*isRunBefore=*/true);
..
createDispatchRegions
..
convertToFlowTensorOps(/*isRunBefore=*/false);
```

Reply:
Yup, I understand that part that we can put them into the same pass and have flags controlling that (though I think it's better to have two passes as we don't share anything here and I think we generally don't want .. flags ;-P). That's a minor issue to me. My main point is that we still need to *identify* and *exclude* the standalone linalg.fill ops in DispatchLinalgOnTensors to make sure it can fall through to be handled by the second invocation of the ConvertToFlowTensorOps pass; so we are gonna have the check in DispatchLinalgOnTensors anyway..

Reply:
> Yup, I understand that part that we can put them into the same pass and have flags controlling that (though I think it's better to have two passes as we don't share anything here and I think we generally don't want .. flags ;-P). That's a minor issue to me.

Acknowledge that its a minor issue, but dont see a point in having all the boiler plate code and two places in the code where conversion happens. Its a flag that is controlled from the pass pipeline in C++ code, not a flag to add to benchmarking pipeline to conditionally use some feature for performance on a particular benchmark. So all I am saying is flags are of different flavors.

> My main point is that we still need to _identify_ and _exclude_ the standalone linalg.fill ops in DispatchLinalgOnTensors to make sure it can fall through to be handled by the second invocation of the ConvertToFlowTensorOps pass; so we are gonna have the check in DispatchLinalgOnTensors anyway..

Good point, that is easy. Just need to add `linalg::FillOp` here https://github.com/google/iree/blob/main/iree/compiler/Dialect/Flow/Transforms/DispatchLinalgOnTensors.cpp#L234.  It will be treated as an op that is always inlined into the dispatch region. If anything is left out, it will be picked up the second invocation of `ConvertToFlow`

Reply:
> Good point, that is easy. Just need to add `linalg::FillOp` here https://github.com/google/iree/blob/main/iree/compiler/Dialect/Flow/Transforms/DispatchLinalgOnTensors.cpp#L234. It will be treated as an op that is always inlined into the dispatch region. If anything is left out, it will be picked up the second invocation of `ConvertToFlow`

I'm not sure that would work. IIUC, it will force fusing the `linalg.fill` with its consumers, even for the cases where we shouldn't, like, it will generate the following:

```mlir
flow.dispatch.workgroups {
  %0 = linalg.fill ... : tensor<1x225x225x3xf32>
  %1 = subtensor_insert %input into %0[0, 0, 0, 0] [1, 224, 224, 3] [1, 1, 1, 1]
        : tensor<1x224x224x3xf32> into tensor<1x225x225x3xf32>
}
```

(A `linalg.fill` and then a `subtensor_insert` into a subrange of it is actually the exact motivating pattern for this change.)

The above fusion is problematic because at the moment we use `linalg.copy` for  `subtensor_insert` and the above dispatch region then will contain multiple linalg ops with different problem sizes, which will mess up distribution. Also after fusing them, there won't exist a standalone `linalg.fill` so the ConvertToFlowTensorOps after DispatchLinalgOnTensors can pick up.. So to make it work, more special checks in `DispatchLinalgOnTensors` pass or even tweaks across the whole pipeline will need to happen. That seems to me even complicated than just 0cc653f40cb49b5b4233a8fe0ef892f42f1c05e7.

Actually coming to your original point here

> adding it by default here (when run before DispatchLinalgOnTensors) will avoid any fusion of fill with other ops.

I'm not sure I follow exactly. The implementation in 91d46f42e47c1e5b3975d26beab35d94bd8221ab checks that we don't have linalg users. If that's true, there won't be any fusion opportunities left for it (as fusion relies on Linalg ops). Am I missing something here?